crates/sui-framework/sources/borrow.move

```diff
@@ -0,0 +1,124 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// A simple library that enables hot-potato-locked borrow mechanics.
+///
+/// With Programmable transactions, it is possible to borrow a value within
+/// a transaction, use it and put back in the end. Hot-potato `Borrow` makes
+/// sure the object is returned and was not swapped for another one.
+module sui::borrow {
+    use sui::object::{Self, ID};
+    use std::option::{Self, Option};
+    use sui::tx_context::TxContext;
+
+    /// The `Borrow` does not match the `Referent`.
+    const EWrongBorrow: u64 = 0;
+    /// An attempt to swap the `Referent.value` with another object of the same type.
+    const EWrongValue: u64 = 1;
+
+    /// An object wrapping a `T` and providing the borrow API.
+    struct Referent<T: key + store> has store {
+        id: ID,
+        value: Option<T>
+    }
+
+    /// A hot potato making sure the object is put back once borrowed.
+    struct Borrow { ref: ID, obj: ID }
+
+    /// Create a new `Referent` struct
+    public fun new<T: key + store>(value: T, ctx: &mut TxContext): Referent<T> {
+        let uid = object::new(ctx);
+        let id = object::uid_to_inner(&uid);
+        let value = option::some(value);
+
+        object::delete(uid);
```

Comment:
Could you use the `value`'s ID instead of generating a new one?  It also means the `Borrow` only needs to store one ID as well.